### ISSUE TITLE
OPDS: Komga CBZ saved as .fb2.zip

### ISSUE THREAD

**Issue by SqueezedByte** [koreader/koreader#7995]
Hi, 
With the latest version of koreader for Kobo if I try to download a .cbz file from the Komga OPDS catalog it will be saved on device as .FB2.ZIP instead of .cbz. so I must rename said file as .cbz to read it.
Is possible to make Koreader save it as .cbz ?

Thanks in advance

**Comment by Frenzie** [comment]
That sounds like it's a problem with the site (i.e., claiming it's something it's not) but could you please provide a link to the website?

**Comment by SqueezedByte** [comment]
Komga Is a selfhosted server a la calibre-web but for comics so I can't give you a link.
This is the project page
https://github.com/gotson/komga

**Comment by Frenzie** [comment]
There's a demo, see https://komga.org/guides/#demo

tl;dr:
```
https://demo.komga.org/opds/v1.2/catalog
demo@komga.org
komga-demo
```

It seems to work. I browsed around and downloaded a file.

![Screenshot_2021-08-22_09-52-39](https://user-images.githubusercontent.com/202757/130346992-047cf828-4ad2-4f1e-a873-88c2507fe87c.png)

(Said file crashes the program @poire-z so attaching it here [Unknown Author - Space_Adventures_002__c2c__diff_ver.fb2.zip](https://github.com/koreader/koreader/files/7026761/Unknown.Author.-.Space_Adventures_002__c2c__diff_ver.fb2.zip), but that aside.)

**Comment by hius07** [comment]
> Said file crashes the program

It is cbz. Why is it named fb2.zip?

**Comment by Frenzie** [comment]
The point is the program should handle it more gracefully. ;-)

It's presumably saved with that extension because: `content-type: application/zip`.

**Comment by Frenzie** [comment]
See https://github.com/koreader/koreader/issues/7218#issuecomment-770423459

**Comment by hius07** [comment]
I renamed (other) fb2 to pdf and it cannot be open ("No reader engine or invalid file").
Isn't it a hard task for the program to distinguish files based on their content, not exention?

**Comment by Frenzie** [comment]
> I renamed (other) fb2 to pdf and it cannot be open ("No reader engine or invalid file").

That's the intended behavior, as opposed to crashing.

**Comment by poire-z** [comment]
> Said file crashes the program

_crash.log, or it didn't happen_:tm: :)

I get this:
08/22/21-10:49:07 DEBUG CreDocument: loading document...
08/22/21-10:49:07 WARN  Cre error loading file: File with supported extension not found in archive.
08/22/21-10:49:07 DEBUG CreDocument: loading failed.
08/22/21-10:49:07 WARN  crengine failed recognizing or parsing this file: unsupported or invalid document

but no crash, just a popup "Failed recognizing or parsing this file... KOReader will exit now." - which is the best we can do in that case (since a long time).
Is this what you get ?

**Comment by Frenzie** [comment]
No, I mean crash.

```
./luajit: frontend/apps/reader/readerui.lua:804: attempt to yield across C-call boundary
stack traceback:
	[C]: in function 'yield'
	frontend/apps/reader/readerui.lua:804: in function 'dealWithLoadDocumentFailure'
	frontend/apps/reader/readerui.lua:285: in function 'v'
	frontend/apps/reader/readerui.lua:434: in function 'init'
	frontend/ui/widget/widget.lua:48: in function 'new'
	frontend/apps/reader/readerui.lua:624: in function 'doShowReader'
	frontend/device/sdl/device.lua:233: in function 'handleSdlEv'
	frontend/device/input.lua:1186: in function 'waitEvent'
	frontend/ui/uimanager.lua:1646: in function 'handleInput'
	frontend/ui/uimanager.lua:1699: in function 'run'
	./reader.lua:323: in main chunk
	[C]: at 0x55784ed8d7a2
~/src/kobo/koreader
```

**Comment by Frenzie** [comment]
@poire-z Did you try on Kobo? Maybe it's SDL-specific.

**Comment by poire-z** [comment]
Just tried it on my Kobo, same behaviour I get on the emulator: popup, KOReader exits - and I don't see your traceback in my crash.log.
Do you get to see the popup I get "...KOreader will exit now" ? In which case it crashes where it should just exists cleanly, so no big deal :)
_attempt to yield across C-call boundary_ is usually something I got when playing with subroutines ....

**Comment by Frenzie** [comment]
> Do you get to see the popup I get "...KOreader will exit now" ? In which case it crashes where it should just exists cleanly, so no big deal :)

No, but you do get that popup with the file manager. Thanks, I didn't think to check that. So it's an edge case related to less usual methods of opening the file, such as "read now" in the OPDS catalog or dragging the file into the window.

**Comment by NiLuJe** [comment]
That check is fishy as hell.

It's assigning the result of `coroutine.running` to a **local** `_coroutine` (while the rest of the module stores that in `self._coroutine`), and fails to actually check that.

Instead, it checks for `coroutine`, which will always be `true`, as that's the `coroutine` module... So it's quite likely potentially yielding when it really shouldn't...

**Comment by NiLuJe** [comment]
As I'm unsure whether we really should stomp/fallback on `self._coroutine` here, see if this helps:

```diff
diff --git a/frontend/apps/reader/readerui.lua b/frontend/apps/reader/readerui.lua
index 05305ab52..21c14b4ff 100644
--- a/frontend/apps/reader/readerui.lua
+++ b/frontend/apps/reader/readerui.lua
@@ -793,7 +793,7 @@ function ReaderUI:dealWithLoadDocumentFailure()
     require("readhistory"):removeItemByPath(self.document.file) -- (will update "lastfile")
     -- As we are in a coroutine, we can pause and show an InfoMessage before exiting
     local _coroutine = coroutine.running()
-    if coroutine then
+    if _coroutine then
         logger.warn("crengine failed recognizing or parsing this file: unsupported or invalid document")
         UIManager:show(InfoMessage:new{
             text = _("Failed recognizing or parsing this file: unsupported or invalid document.\nKOReader will exit now."),
```

**Comment by poire-z** [comment]
^ Yup, that coroutine missing its _ looks like a typo of mine :)

But it may have shown that _less usual methods of opening the file, such as "read now" in the OPDS catalog or dragging the file into the window_ may not be run in the coroutine ReaderUI is usually run in - and won't show this message and exit (just as with the crash), so there may be something else to clean up/make consistent.
